### ISSUE TITLE
[Core]  Persist Ray runtime resources between sessions

### ISSUE THREAD

**Issue by joncarter1** [ray-project/ray#37377]
### Description

Currently Ray runtime resources are cached at:
`TEMP_DIR/SESSION_DIR/runtime_resources`
where TEMP_DIR can be set with an environment variable e.g. `ray start --temp-dir /alt/dir`
and:
`SESSION_DIR = session_{date_str}_{os.getpid()}` is a fixed unique hash based on the Ray start time.
e.g. 
`/tmp/dir/session_2023-07-11_14-35-32_145288_101538/runtime_resources`

AFAICS there is no way to persist runtime_resources between Ray instances i.e. when Ray is restarted.

Is there currently any way to do this? (Hacky suggestions welcome)

If not, perhaps this could be solved by making the resources directory or the session directory overridable.

**Comment by rkooo567** [comment]
Can you describe the use case in more details?

**Comment by joncarter1** [comment]
Sure.

I must admit my scenario probably isn't that common.
The TLDR is I've managed to hack something together with a custom RuntimeEnv plugin.

## Set-up
I share a collection of machines with other users.
I use Docker containers on each machine to form a Ray cluster. These are scheduled using Hashicorp Nomad.
I automatically spin up and shut down the Ray cluster around each job, for effective resource sharing.

## Problem
To ship my (Conda) dependencies to the cluster I see a few existing options:
#### 1. Re-install the Conda environment from the RuntimeEnv in each new cluster instance.
I'd rather not do this - the Conda environments used are bulky and take 5-10 minutes to install (even with extensions like libmamba). Obviously this isn't an issue with e.g. multi-day GPU training, but it is with some of my other use-cases, where each job takes 30mins to an hour.

#### 2. Add the Conda environments to the Docker images.
Then reference the environments by name in the RuntimeEnv.
I'd like to avoid this approach to spend less time building and shipping containers new containers as and when dependencies change.

Instead, what I'd like to be able to do is still leverage the RuntimeEnv, but have the cache of runtime_resources (particularly Conda environments) persist between Ray instances, rather than creating a new directory every time. 

## My solution
I've managed to solve this with a RuntimeEnv plugin that sub-classes the existing CondaPlugin.

This looks something like:
https://gist.github.com/joncarter1/3d9d21eb1b20673ceca7687cb1a484b5

## To the issue
Your call as to whether a configurable and persistent runtime_resources directory is something worth supporting natively.
 I am not blocked by this, just a nice-to-have. :smile:

**Comment by joncarter1** [comment]
Closing this issue because \*\*joins the back of the queue\*\* uv has made my problems go away.

God bless @astral-sh and @charliermarsh 🙌

**Comment by charliermarsh** [comment]
Hah, that made me laugh, thanks @joncarter1! Glad uv is helping here.
